core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java

```diff
@@ -421,4 +512,29 @@ public void noMoreSinkFactories()
             exchange.noMoreSinkFactories();
         }
     }
+
+    private static class ProjectingPartitionFunction
+            implements PartitionFunction
+    {
+        private final PartitionFunction delegate;
+        private final int[] partitionChannels;
+
+        private ProjectingPartitionFunction(PartitionFunction partitionFunction, int[] partitionChannels)
+        {
+            this.delegate = requireNonNull(partitionFunction, "partitionFunction is null");
+            this.partitionChannels = requireNonNull(partitionChannels, "partitionChannels is null");
+        }
+
+        @Override
+        public int getPartitionCount()
+        {
+            return delegate.getPartitionCount();
+        }
+
+        @Override
+        public int getPartition(Page page, int position)
+        {
+            return delegate.getPartition(page.getColumns(partitionChannels), position);
```

Comment:
Calling `page.getColumns(partitionChannels)` once per position row seems like it would be pretty expensive.

Reply:
Good catch: https://github.com/trinodb/trino/pull/7112